Manatee.Json/Serialization/Internal/Serializers/JsonValueSerializer.cs

```diff
@@ -0,0 +1,154 @@
+using System;
+using System.Linq;
+using JetBrains.Annotations;
+
+namespace Manatee.Json.Serialization.Internal.Serializers
+{
+    [UsedImplicitly]
+    internal class JsonValueSerializer : IPrioritizedSerializer
+    {
+        public int Priority => 2;
+        public bool ShouldMaintainReferences => true;
+
+        private Type[] HandledTypes = {
+            typeof(JsonValue),
+            typeof(JsonObject),
+            typeof(JsonArray)
+        };
+
+        public bool Handles(SerializationContextBase context)
+        {
+            return HandledTypes.Contains(context.InferredType);
+        }
+
+        public JsonValue Serialize(SerializationContext context)
+        {
+            switch (context.Source)
+            {
+                case JsonValue jsonValue:
+                    switch(jsonValue.Type)
+                    {
+                        case JsonValueType.Object:
+                            return SerializeJsonObject(context, jsonValue.Object);
+                        case JsonValueType.Array:
+                            return SerializeJsonArray(context, jsonValue.Array);
+                        case JsonValueType.Number:
+                        case JsonValueType.String:
+                        case JsonValueType.Boolean:
+                        case JsonValueType.Null:
+                        default:
+                            return jsonValue;
+                    }
+                case JsonObject jsonObject:
+                    return SerializeJsonObject(context, jsonObject);
+                case JsonArray jsonArray:
+                    return SerializeJsonArray(context, jsonArray);
+                default:
+					throw new ArgumentOutOfRangeException();
+            }
+        }
+
+        private JsonValue SerializeJsonObject(SerializationContext context, JsonObject jsonObject)
+        {
+            var result = new JsonObject();
+            foreach (var kvp in jsonObject)
+            {
+                var inferredValue = GetInferredValue(kvp.Value);
+                context.Push(inferredValue.GetType(), typeof(JsonValue), kvp.Key, inferredValue);
+                result.Add(kvp.Key, context.RootSerializer.Serialize(context));
+                context.Pop();
+            }
+            return result;
+        }
+
+        private JsonValue SerializeJsonArray(SerializationContext context, JsonArray jsonArray)
+        {
+            return jsonArray.Select((value, i) =>
+            {
+                var inferredValue = GetInferredValue(value);
+                context.Push(inferredValue.GetType(), typeof(JsonValue), i.ToString(), inferredValue);
+                var json = context.RootSerializer.Serialize(context);
+                context.Pop();
+                return json;
+            }).ToList().ToJson();
+        }
+
+        public object Deserialize(DeserializationContext context)
+        {
+            switch (context.LocalValue.Type)
```

Comment:
Here you need to check the type that the serializer is requesting over the type that you're given.

Reply:
I'm not sure if I follow. Is it not such that when deserializing from any to `JsonValue`, `context.RequestedType` is always `JsonValue`?

At least, when I stepped through the code, the requested type was always found to be `JsonValue`.

I could for instance add:
```
if (context.RequestedType != typeof(JsonValue))
{
    throw new NotImplementedException();
}
```

Reply:
Yeah, saying that is what I'm looking for.  The purpose is to deserialize into what the actual model is asking for, not what the serializer _thinks_ should be there.

Think about the case where the property is `JsonArray`.  If you give it a `JsonValue`, it'll fail.